File: benchmarks/python/hashmap.py
#!/usr/bin/env python3

# Hash Map benchmark in Python

class HashMap:
    def __init__(self, capacity=16):
        self.capacity = capacity
        self.buckets = [[] for _ in range(capacity)]
        self.size = 0
    
    def _hash(self, key):
        return hash(key) % self.capacity
    
    def put(self, key, value):
        index = self._hash(key)
        bucket = self.buckets[index]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)
                return
        bucket.append((key, value))
        self.size += 1
    
    def get(self, key):
        index = self._hash(key)
        bucket = self.buckets[index]
        for k, v in bucket:
            if k == key:
                return v
        return None
    
    def remove(self, key):
        index = self._hash(key)
        bucket = self.buckets[index]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                del bucket[i]
                self.size -= 1
                return True
        return False
    
    def contains(self, key):
        return self.get(key) is not None

# Create and populate hash map
hm = HashMap()
for i in range(5000):
    hm.put(f"key_{i}", i * 2)

# Get values
total = 0
for i in range(0, 5000, 10):
    val = hm.get(f"key_{i}")
    if val is not None:
        total += val

# Remove some
for i in range(0, 2500, 10):
    hm.remove(f"key_{i}")

print(f"Map size: {hm.size}, Sum: {total}")

File: benchmarks/javascript/hashmap.js
// Hash Map benchmark in JavaScript

class HashMap {
    constructor(capacity = 16) {
        this.capacity = capacity;
        this.buckets = new Array(capacity).fill(null).map(() => []);
        this.size = 0;
    }
    
    _hash(key) {
        let hash = 0;
        const str = String(key);
        for (let i = 0; i < str.length; i++) {
            hash = ((hash << 5) - hash) + str.charCodeAt(i);
            hash = hash & hash;
        }
        return Math.abs(hash) % this.capacity;
    }
    
    put(key, value) {
        const index = this._hash(key);
        const bucket = this.buckets[index];
        for (let i = 0; i < bucket.length; i++) {
            if (bucket[i][0] === key) {
                bucket[i] = [key, value];
                return;
            }
        }
        bucket.push([key, value]);
        this.size++;
    }
    
    get(key) {
        const index = this._hash(key);
        const bucket = this.buckets[index];
        for (const [k, v] of bucket) {
            if (k === key) {
                return v;
            }
        }
        return undefined;
    }
    
    remove(key) {
        const index = this._hash(key);
        const bucket = this.buckets[index];
        for (let i = 0; i < bucket.length; i++) {
            if (bucket[i][0] === key) {
                bucket.splice(i, 1);
                this.size--;
                return true;
            }
        }
        return false;
    }
    
    contains(key) {
        return this.get(key) !== undefined;
    }
}

// Create and populate hash map
const hm = new HashMap();
for (let i = 0; i < 5000; i++) {
    hm.put(`key_${i}`, i * 2);
}

// Get values
let total = 0;
for (let i = 0; i < 5000; i += 10) {
    const val = hm.get(`key_${i}`);
    if (val !== undefined) {
        total += val;
    }
}

// Remove some
for (let i = 0; i < 2500; i += 10) {
    hm.remove(`key_${i}`);
}

console.log(`Map size: ${hm.size}, Sum: ${total}`);
